### ISSUE TITLE
Scaling factor API changes break existing image edits

### ISSUE THREAD

**Issue by jakubfi** [lensfun/lensfun#945]
Scale factor used by lfModifier::EnableScaling() and lfModifier::GetAutoScale() has different meaing in 0.3.2 and git lensfun versions. This results in broken image edits done in a program compiled with stable lensfun version when opened in a program compiled with current development lensfun version.

Actual case: When lens correction is applied in darktable compiled against lensfun-0.3.2 sidecar file is written with, say, scaling factor set to 1.033 (as computed by the library). Opening the same edit in darktable compiled against lensfun-git sets the scaling factor to 1.033 again, causing distortion correction artifacts on the edge of the image. When lens corrections are re-applied, scaling factor is now set to 1.047 and image looks good again.

Automatically resetting the scale factor in darktable after switching to lensfun-next from 0.3.2 is not a solution, because user may set have it to a deifferent, preferred value (which API allows for).

lensfun version should not redefine the actual meaning of the scaling factor (even if it was wrong in the beginning). Instead, it should do all the necessary back and forth calculations behind the scenes of EnableScaling() and GetAutoScale(), and keep the scaling API consistent across versions.

**Comment by seebk** [comment]
Not sure if we can convert the scale factor or if it makes sense. Maybe @bronger has an idea?

I will put it to the next release milestone.

**Comment by jakubfi** [comment]
@seebk @bronger If converting the scale factor behind the scenes is not feasible, maybe provide a method to allow external applications to convert it when necessary? I don't know all the lensfun usage scenarios, but I would assume that darktable is not the only one to bump into this in the future.

**Comment by uchrisu** [comment]
Okay, it seems to be due to the "ru" stretching as distortion parameters are converted to keep the real focal length. "ru" is internally stretched by factor "d". I would suggest to add member d in lfLensCalibDistortion wehn transforming the distortion coefficients. Then GetAutoScale and EnableScaling can use this "d" factor by simply dividing/multiplying the d factor to get back to the former scaling factor. 

I will implement this and check the results against the results from 0.3.92 if no one complains about this way.

**Comment by uchrisu** [comment]
However, changing back the meaning of the scale we loose the convention that scale=1 preserves the real focal length. 

Another problem: Many people are already using 0.3.95 at the moment. When we bring back the scale interpretation as in 0.3.92 pictures get broken for them. 

Maybe we should keep it as it is now? Or should we add at least functions like GetREALAutoScale and EnableREALScaling (without all Caps, of course)?

**Comment by seebk** [comment]
My two cents:

- Lensfun should maintain compatibility to 0.3.2 but not to any 0.3.9x versions. The latter are preview versions and not meant for productive use.
- Passing additional abstract internal parameters like `d` to `EnableScaling` is IMO not a good idea. The public API should not expose internal details.
- If necessary, adding an additional `EnableScaling...()` method to achieve backwards compatibility might be a good workaround. It still keeps the new API with the correct(?) definition of the scale factor but allows old programs to be backwards compatible. However, we have to consider the "enabled flags" have to be extended, too. 
- Are we sure that the new scale factor definition is correct or better than the previous one in 0.3.2? I am not too familiar with the math and physics behind Lensfun. Might be worth to wait for feedback from @bronger on that topic...

**Comment by seebk** [comment]
Another idea... is it possible to create a `ConvertScaleFactor()` function that Darktable can use to convert the old scale factor to the new one?

**Comment by uchrisu** [comment]
>     * Passing additional abstract internal parameters like `d` to `EnableScaling` is IMO not a good idea. The public API should not expose internal details.

Sorry, maybe my suggestion was unclear. What I really meant: d can be an "private" member parameter in the struct lfLensCalibDistortion. It can be set when the parameters (e.g., a,b,c for PTLens) are rescaled and GetAutoScale and EnableScaling can use this "internally" without the knowledge of the API-user. 



>     * If necessary, adding an additional `EnableScaling...()` method to achieve backwards compatibility might be a good workaround. It still keeps the new API with the correct(?) definition of the scale factor but allows old programs to be backwards compatible. However, we have to consider the "enabled flags" have to be extended, too.

Yes either this way or vise versa. The other way would be to introduce new names for the functions with the scaling factors as implemented in the master now.

**Comment by seebk** [comment]
Ok, got it. If adding the d parameter to the lfLensCalibDistortion struct is sufficient it sounds like the best option. But how does the `ModifyCoord_Scale()` get access to the `d` parameter from `lfLensCalibDistortion`? Anyway, if you see a clear path to implement it feel free to get it started :) Still, I'd propose to await @bronger's feedback before merging it.

**Comment by uchrisu** [comment]
The best way was to put d as private member in lfModifier.
Here is the fix: #1174

**Comment by payano** [comment]
Hello @uchrisu, I think that I'm kind of late to party, but is there a way that we can fix this issue some how?

**Comment by seebk** [comment]
@payano I think the latest modified patch by @uchrisu was working but needs final review and merging. See the history of the discussion here and in the PR #1174. Unfortunateley, there currently is no active core developer in the Lensfun team so this might still take some time until it finally is getting merged.

**Comment by payano** [comment]
@seebk thanks for the information, it's unfortunate if there's a solution but it hasn't been merged :(

**Comment by bronger** [comment]
> […]
> 
> Actual case: When lens correction is applied in darktable compiled against lensfun-0.3.2 sidecar file is written with, say, scaling factor set to 1.033 (as computed by the library). Opening the same edit in darktable compiled against lensfun-git sets the scaling factor to 1.033 again, causing distortion correction artifacts on the edge of the image. When lens corrections are re-applied, scaling factor is now set to 1.047 and image looks good again.

What does “distortion correction artifacts” mean? Is there residual distortion, or are there black blotches at the border?

**Comment by seebk** [comment]
Yes, I think this is what we see in other applications as black borders. In Darktable pixels outside of the original image are not drawn black but clamped to the closest pixel on the image edge leading to these strange artefacts as you can see here

<img width="720" height="806" alt="Image" src="https://github.com/user-attachments/assets/bdf171f2-e38a-45e8-ade8-6605f617ff6f" />
